unison-share-api/src/Unison/Server/Endpoints/NamespaceListing.hs

```diff
@@ -154,61 +159,92 @@ serve ::
-      namespaceListing = do
-        -- Relative and Listing Path resolution
-        --
-        -- The full listing path is a combination of the relativeToPath (prefix) and the namespace path
-        --
-        -- For example:
-        --            "base.List"    <>    "Nonempty"
-        --                ↑                    ↑
-        --         relativeToPath        namespacePath
-        --
-        -- resulting in "base.List.map" which we can use via the root branch (usually the codebase hash)
-        -- to look up the namespace listing and present shallow name, so that the
-        -- definition "base.List.Nonempty.map", simple has the name "map"
-        --
-        relativeToPath' <- (parsePath . Text.unpack) $ fromMaybe "." mayRelativeTo
-        namespacePath' <- (parsePath . Text.unpack) $ fromMaybe "." mayNamespaceName
-
-        let path = Path.fromPath' relativeToPath' <> Path.fromPath' namespacePath'
-        let path' = Path.toPath' path
-
-        mayRootHash <- traverse (Backend.expandShortBranchHash codebase) maySBH
-        listingCausal <- Backend.getShallowCausalAtPathFromRootHash codebase mayRootHash path
-        listingBranch <- liftIO $ V2Causal.value listingCausal
-        shallowPPE <- liftIO $ Backend.shallowPPE codebase listingBranch
-        let listingFQN = Path.toText . Path.unabsolute . either id (Path.Absolute . Path.unrelative) $ Path.unPath' path'
-        let listingHash = v2CausalBranchToUnisonHash listingCausal
-        listingEntries <- lift (findShallow listingBranch)
-
-        makeNamespaceListing shallowPPE listingFQN listingHash listingEntries
-   in namespaceListing
+serve codebase maySBH mayRelativeTo mayNamespaceName = do
+  useIndex <- asks Backend.useNamesIndex
+  mayRootHash <- traverse (Backend.expandShortBranchHash codebase) maySBH
+  codebaseRootHash <- liftIO $ Codebase.getRootBranchHash codebase
+
+  --Relative and Listing Path resolution
+  --
+  -- The full listing path is a combination of the relativeToPath (prefix) and the namespace path
+  --
+  -- For example:
+  --            "base.List"    <>    "Nonempty"
+  --                ↑                    ↑
+  --         relativeToPath        namespacePath
+  --
+  -- resulting in "base.List.map" which we can use via the root branch (usually the codebase hash)
+  -- to look up the namespace listing and present shallow name, so that the
+  -- definition "base.List.Nonempty.map", simple has the name "map"
+  --
+  relativeToPath' <- (parsePath . Text.unpack) $ fromMaybe "." mayRelativeTo
+  namespacePath' <- (parsePath . Text.unpack) $ fromMaybe "." mayNamespaceName
+
+  let path = Path.fromPath' relativeToPath' <> Path.fromPath' namespacePath'
+  let path' = Path.toPath' path
+
+  case (useIndex, mayRootHash) of
+    (True, Nothing) ->
+      serveFromIndex codebase mayRootHash path'
+    (True, Just rh)
+      | rh == causalHash2to1 codebaseRootHash ->
+        serveFromIndex codebase mayRootHash path'
+    (_, Just rh) -> do
```

Comment:
```suggestion
    (False, Just rh) -> do
```

Reply:
It's not equivalent, since this also catches the fall-through case if `rh /= causalHash2to1 codebaseRootHash`; but I can add an extra case to make both of those explicit if you like